feat(Analysis/Asymptotics): define subpolynomial growth

0xTerencePrime opened the pull request:
## Main definitions

* `Asymptotics.IsSubpolynomial l f g`: A function `f` has subpolynomial growth with respect to `g` along filter `l` if `f = O(1 + ‖g‖^k)` for some natural `k`.

## Main results

* `IsSubpolynomial.const`: Constant functions have subpolynomial growth
* `IsSubpolynomial.id`: Identity has subpolynomial growth  
* `IsSubpolynomial.add`: Closure under addition
* `IsSubpolynomial.neg`: Closure under negation
* `IsSubpolynomial.sub`: Closure under subtraction
* `IsSubpolynomial.mul`: Closure under multiplication
* `IsSubpolynomial.pow`: Closure under powers
* `isSubpolynomial_iff_one_add`: Equivalence with `(1 + ‖g‖)^k` formulation
* `IsSubpolynomial.uniform`: Uniform bounds for finite families

## Implementation notes

The definition uses `1 + ‖g‖^k` rather than `(1 + ‖g‖)^k` as the primary form, with the equivalence established in `isSubpolynomial_iff_one_add`. Four private auxiliary lemmas handle the key inequalities needed for closure proofs.

Closes #32658

0xTerencePrime committed:
feat(Analysis/Asymptotics): define subpolynomial growth

- Define `IsSubpolynomial` for functions with subpolynomial growth
- Prove stability under addition, multiplication, and powers
- Prove equivalence with alternative definition using (1 + ‖g‖)^k
- Prove uniform bound for finite families

Closes #32658

0xTerencePrime committed:
style: fix trailing newline in Subpolynomial.lean

0xTerencePrime committed:
fix: mark declarations public to satisfy linter

0xTerencePrime committed:
Merge branch 'master' into feat/subpolynomial-growth

ADedecker reviewed (approved):
Thanks for your work!

I'm sorry for leaving this aside for so long, I was away from my computer for the holidays. I will give this a closer look in the following days, but here is a first general comment:

I think I made your life a lot more painful than necessary by suggesting to use `1 + ‖g‖^k` instead of `(1 + ‖g‖)^k`. For some reason I thought this was more natural, but the pain you had to go through shows that `(1 + ‖g‖)^k` is better because `1 + ‖g‖ ≥ 1` so you can avoid a lot of precise computations (e.g, if `k ≤ n`, one directly has `(1 + ‖g‖)^k ≤ (1 + ‖g‖)^n`).

In the end both do end up being equivalent, but I think that the theory should be developed using the more convenient form; would you mind refactoring your code to fit with this? Sorry again for the extra trouble, I definitely should have thought a bit more while writing the issue...

0xTerencePrime committed:
docs: update author and remove issue reference

0xTerencePrime committed:
refactor: use (1 + ‖g‖)^k as primary definition

0xTerencePrime committed:
Merge branch 'master' into feat/subpolynomial-growth

0xTerencePrime commented on Mathlib/Analysis/Asymptotics/Subpolynomial.lean:
```diff
@@ -0,0 +1,288 @@
+/-
+Copyright (c) 2024 0xTerencePrime. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: 0xTerencePrime
```
Done.

0xTerencePrime commented on Mathlib/Analysis/Asymptotics/Subpolynomial.lean:
```diff
@@ -0,0 +1,288 @@
+
+## References
+
+* Issue: https://github.com/leanprover-community/mathlib4/issues/32658
```
Removed.

0xTerencePrime commented:
Refactored to use (1 + ‖g‖)^k as suggested. The proofs are now significantly simpler.

-awaiting-author

ADedecker commented on Mathlib/Analysis/Asymptotics/Subpolynomial.lean:
```diff
@@ -0,0 +1,220 @@
+  exact isBigO_const_const c (by norm_num : (1 : ℝ) ≠ 0) l
+
+/-- The function `x ↦ ‖g x‖` has subpolynomial growth with respect to `g`. -/
+public theorem IsSubpolynomial.id (l : Filter α) (g : α → F) :
```
I think this should be called `IsSubpolynomial.self`, and be stated as `IsSubpolynomial l g g`.

Regarding `g` vs `fun x => ‖g x‖`, please add some lemmas analogous to [Asymptotics.isBigO_norm_left](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Asymptotics/Defs.html#Asymptotics.isBigO_norm_left) and [Asymptotics.isBigO_norm_right](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Asymptotics/Defs.html#Asymptotics.isBigO_norm_right)

ADedecker commented on Mathlib/Analysis/Asymptotics/Subpolynomial.lean:
```diff
@@ -0,0 +1,220 @@
+  simp only [Real.norm_eq_abs, one_mul]
+  rw [abs_of_nonneg (norm_nonneg _)]
+  rw [abs_of_pos (by linarith [norm_nonneg (g x)] : 0 < 1 + ‖g x‖)]
+  linarith [norm_nonneg (g x)]
```
```suggestion
  apply IsBigO.of_norm_le
  simp [pow_one]
```

ADedecker reviewed (commented):
Here is a first round of comments on the new version. These are quite heavy golfs, but keep in mind that these ideas come from 1) experience and 2) looking at similar things in the library. For example, seeing [Asymptotics.isBigO_iff_eventually_isBigOWith](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Asymptotics/Defs.html#Asymptotics.isBigO_iff_eventually_isBigOWith) pop up in some existing proof made me realize that we should use something similar here.

A global remark: all of your lemmas include some filter `l`, and most of them also talk about the two functions `f` and `g`. Could you move them to `variables`? Note that you can still make them explicit by doing `variable (l) in` before a theorem/def.

Do not hesitate to ask for more details and explanations!

0xTerencePrime committed:
refactor: simplify proofs using filter_upwards and variables

0xTerencePrime commented on Mathlib/Analysis/Asymptotics/Subpolynomial.lean:
```diff
@@ -0,0 +1,220 @@
+* `IsSubpolynomial.id`: The norm of the reference function is subpolynomial.
+-/
+
+open Filter Asymptotics
```
Done.

0xTerencePrime commented on Mathlib/Analysis/Asymptotics/Subpolynomial.lean:
```diff
@@ -0,0 +1,220 @@
+    rw [abs_of_pos (pow_pos (by linarith [norm_nonneg (g x)]) k₂)]
+    rw [abs_of_pos (pow_pos (by linarith [norm_nonneg (g x)]) k)]
+    exact pow_mono_height (norm_nonneg _) (le_max_right _ _)
+  exact (hk₁.trans h1).add (hk₂.trans h2)
```
Done. Added isSubpolynomial_iff_eventually_isBigO and simplified the proof.

0xTerencePrime commented on Mathlib/Analysis/Asymptotics/Subpolynomial.lean:
```diff
@@ -0,0 +1,220 @@
+
+/-- For a finite family of subpolynomial functions, one can choose a uniform degree and constant. -/
+public theorem IsSubpolynomial.uniform {ι : Type*} {s : Finset ι} {l : Filter α}
+    {f : ι → α → E} {g : α → F}
```
Done.

0xTerencePrime committed:
chore: address remaining formatting and API review feedback

0xTerencePrime committed:
chore: merge master to update dependencies and fix CI

0xTerencePrime committed:
fix: close missing section to resolve CI warning

0xTerencePrime committed:
fix: restore module and public keywords for Lean 4 module system compatibility

0xTerencePrime commented:
Sorry for being MIA these last few months! Things got pretty hectic for me and I just couldn't find the time to address this until now. thank you for your patience, @ADedecker!

In response to your inquiry about the AI: yes, I employed an AI assistant in order to generate docstrings, format boilerplate code, and learn about new techniques like `bound` and `positivity`. The mathematical reasoning and the proofs themselves are my own work; I have gone over everything to verify that they are mathematically valid and conform to contribution standards for Mathlib.

In the latest commits I have applied all of your suggested changes:
- Simplified the two auxiliary lemmas via the `bound` tactic
- Replaced the `abs_of_pos` rewrites via `simp (discharger := positivity) only [Real.norm_of_nonneg]`
- Updated the section header from "Equivalence with the old definition" to the more descriptive `` `O(1 + ‖g‖^k)` ``
- Removed the unnecessary `public` modifier from `IsSubpolynomial.uniform`
- Merged the latest `master` changes

Now all 9 CI checks are passing. Thanks for your patience and fantastic guidance once more!

-awaiting-review
